### DISCUSSION TITLE
SMILESのEmbeddingsを取得してt-SNE

### DISCUSSION THREAD

**Discussion by YumizSui** [souyakuchan/LLM_DD_Challenge#5] [Show and tell]
OpenAIの`text-embedding-ada-002`モデルを用いて、LLMがどのくらいSMILESを理解できてそうか検証してみた

参考：https://twitter.com/YumizSui/status/1642763480557301760

再現コードは以下

```python
import numpy as np
import pandas as pd
import openai
from rdkit import Chem
from rdkit.Chem import PandasTools
from sklearn.manifold import TSNE
import matplotlib
import matplotlib.pyplot as plt

# Load data
df = PandasTools.LoadSDF("data/CBLB_inhibitors_vsF.sdf")
smiles = df["ROMol"].apply(Chem.MolToSmiles).tolist()

# Define Active/Inactive
strong_actives = ['<100', '101-300']
weak_actives = ['301-1000', '101-1000']
df['active']=df['IC50_range_nM'].apply(lambda x: 2 if x in strong_actives else 1 if x in weak_actives else 0)

# Get embbedings
results = openai.Embedding.create(input=smiles, model="text-embedding-ada-002")
embeddings = np.array([results['data'][i]['embedding'] for i in range(len(smiles))])

# Convert to tsne space
tsne = TSNE(n_components=2, perplexity=30, early_exaggeration=30, random_state=42, init='random', learning_rate=200)
vis_dims = tsne.fit_transform(embeddings)

# Plot
x = [x for x,y in vis_dims]
y = [y for x,y in vis_dims]

color_indices = df['active']
colors = ["darkgreen", "gold", "red"]
colormap = matplotlib.colors.ListedColormap(colors)

plt.scatter(x, y, c=color_indices, cmap=colormap, alpha=0.3)

plt.scatter([],[], color=colormap(0), label='inactive', alpha=0.5)
plt.scatter([],[], color=colormap(1), label='week-active', alpha=0.5)
plt.scatter([],[], color=colormap(2), label='active', alpha=0.5)
plt.legend()
plt.title("Embedded SMILES with text-embedding-ada-002 using t-SNE")
```
以下EmbeddingsとECFP4のt-SNEそれぞれ描画しました。

![image](https://user-images.githubusercontent.com/55099270/229436945-3e6a8e8c-a126-48cb-b181-4dcc3f7efa12.png)
![image](https://user-images.githubusercontent.com/55099270/229436996-256f4593-76ee-458e-bc14-c09f5802954a.png)

ある程度構造を見ることは出来てそう

使えるかは不明だがこちら
[text-embedding-ada-002.csv](https://github.com/souyakuchan/LLM_DD_Challenge/files/11135738/text-embedding-ada-002.csv)

**Comment by YumizSui** [comment]
どうやってt-SNEのパラメータ決めるべきかGPT-4に聞いみた

# yumizsui

How do you think I should set the perplexity in t-sne to convert a 1536 dimensional vector into a 2 dimensional vector with 896 data points?

# GPT-4

T-distributed Stochastic Neighbor Embedding (t-SNE) is a popular dimensionality reduction technique that is particularly well-suited for visualizing high-dimensional data in a lower-dimensional space, such as 2D. Perplexity is a crucial hyperparameter in t-SNE that influences the balance between local and global structures in the resulting visualization.

There is no one-size-fits-all answer for choosing the perplexity value, as the optimal value may vary depending on the structure of your dataset. However, some general guidelines can help you make an informed decision:

1. Start with a moderate perplexity value: Researchers often recommend starting with a perplexity value between 5 and 50. A common default choice is 30.
2. Experiment with different values: To find the best perplexity for your specific dataset, you can try a range of values and examine the resulting visualizations. Assess the quality of the visualizations based on the preservation of local and global structures, as well as the presence of meaningful clusters.
3. Consider the number of data points: In your case, you have 896 data points. It is generally suggested that the perplexity should be smaller than the number of data points. A perplexity value that is too large may lead to overly dense clusters or loss of local structure, while a value that is too small may result in a more fragmented visualization.

Given these guidelines, you could start with a perplexity of 30 and experiment with different values in the range of 5 to 50 (or even up to 100 if you want to explore more extreme cases). Observe the visualizations to determine which value best represents the structure of your data in a 2D space. Keep in mind that t-SNE is sensitive to other hyperparameters as well, such as the learning rate and the number of iterations, so you may want to experiment with those too to optimize your results.

# yumizsui

How do I determine early_exaggeration?

# GPT-4

The early_exaggeration parameter in t-SNE is another important hyperparameter that controls the emphasis on retaining local structure during the optimization process. Essentially, early_exaggeration inflates the distances between data points in the high-dimensional space at the beginning of the optimization, which helps separate clusters more distinctly in the lower-dimensional space.

Here are some guidelines for choosing an appropriate early_exaggeration value:

1. Default value: The default early_exaggeration value in most t-SNE implementations, including the popular scikit-learn library, is 12.0. This value often works well for many datasets, so it's a good starting point.
2. Experiment with different values: Similar to choosing the perplexity value, it's a good idea to experiment with different early_exaggeration values to see which one produces the best results for your dataset. You can try values between 1 and 100, but it's more common to test values in the range of 5 to 50.
3. Observe the impact on cluster separation: When you experiment with different early_exaggeration values, observe how they affect the separation of clusters in the resulting visualization. Higher values of early_exaggeration will lead to more distinct clusters, but they may also result in exaggerated distances between clusters, potentially separating points that should be closer together. On the other hand, smaller values may not separate clusters as effectively, leading to a more uniform distribution of points in the lower-dimensional space.
4. Balance local and global structure: The goal is to find a value for early_exaggeration that balances the preservation of local structure while still maintaining a meaningful global structure. If you observe that increasing early_exaggeration leads to better-defined clusters but deteriorates the global structure, consider using a lower value.

Remember that t-SNE is sensitive to several hyperparameters, so it's essential to experiment with different combinations of perplexity, early_exaggeration, learning rate, and the number of iterations to find the best visualization for your dataset.

**Comment by rkakamilan** [comment]
MoleculeNetのlipophilicity データを使った回帰タスクでopenaiのembedding と既知フィンガープリント、記述子だけでなくChemBERTaなどとも比較したベンチマークが出てました

https://maclandrol.github.io/molfeat-hype/main/tutorials/benchmark.html
